### DISCUSSION TITLE
Alsa crossfeed plugin

### DISCUSSION THREAD

**Discussion by timsofteng** [MusicPlayerDaemon/MPD#1410] [General]
Hello.

I'm using mpd with my external audio interface. I send bitperfect stream from PC to interface with alsa.

Is it any existing way to add crossfeed effect for headphones listening without resempling and another things to keep bitperfect working?

**Comment by MaxKellermann** [comment]
You can have two (or more) audio outputs, and enable/disable them at runtime. So whenever you want to use your headphones, enable the ALSA device where you connected them.

**Reply by timsofteng** [reply]
> You can have two (or more) audio outputs, and enable/disable them at runtime. So whenever you want to use your headphones, enable the ALSA device where you connected them.

Thanks for the reply. How can I do it?
How can I add this plugin?

**Reply by MaxKellermann** [reply]
It's documented here: https://mpd.readthedocs.io/en/stable/user.html#configuring-audio-outputs
But I wonder, if you already set this up, and you know it's bit-perfect, you should know that already, don't you?
What makes you sure your current setup is bit-perfect?

**Reply by timsofteng** [reply]
To be honest I can't be sure it is bit perfrect stream. I've just added this output and it works and sounds pretty good

audio_output {
  type "alsa"
  name "Motu M2"
  device "hw:1,0"

  auto_resample "no"
  auto_channels "no"
  auto_format "no"

  mixer_type "none"
  replay_gain_handler "none"
}

Is `filter` option what I'm looking for?
Unfortunately can't find `crossfeed` there...

**Reply by MaxKellermann** [reply]
"Sounds pretty good" isn't a good benchmark. I bet you wouldn't be able to recognize non-bit-perfect playback in a blind test. I certainly wouldn't.
Your configuration looks correct, even though those `auto_` options are not necessary; they have no effect in `hw:` configurations, because that's ALSA's default already.

But it turns out I misunderstood your initial question. I thought by "crossfeed" you mean to feed both speakers and headphones, but you didn't say that you have *only* headphones.

"Crossfeed" is actually a well-defined term, which I did not know. The simple answer is: no, it is not logically possible to add cross-feed (or any other effect) while at the same time staying bit-perfect. Any effect, no matter which, modifies the audio stream, and modifying the audio stream is the exact opposite of "bit-perfect".

**Reply by MaxKellermann** [reply]
btw. even if your configuration is correct, it still does not imply that you're really bit-perfect; this depends on the capabilities of your sound hardware; see https://mpd.readthedocs.io/en/stable/user.html#bit-perfect-playback for details.

**Reply by timsofteng** [reply]
Thanks for this useful info.
Hm...

Looks like my Motu can only play 32-bit files... Right?

Jan 22 20:37 : playlist: queue song 360:"Black Sabbath - 2016 - Paranoid (Deluxe Edition 2CD)/Disc 1/05 Electric Funeral.flac"
Jan 22 20:37 : decoder_thread: probing plugin flac
Jan 22 20:37 : decoder: audio_format=44100:16:2, seekable=true
Jan 22 20:37 : alsa_output: opened hw:1,0 type=HW
Jan 22 20:37 : alsa_output: buffer: size=16..131072 time=362..2972155
Jan 22 20:37 : alsa_output: period: size=8..65536 time=181..1486078
Jan 22 20:37 : alsa_output: default period_time = buffer_time/4 = 500000/4 = 125000
Jan 22 20:37 : alsa_output: format=S32_LE (Signed 32 bit Little Endian)
Jan 22 20:37 : alsa_output: buffer_size=22050 period_size=5513
Jan 22 20:37 : output: opened "Motu M2" (alsa) audio_format=44100:32:2
Jan 22 20:37 : output: converting in=44100:16:2 -> f=44100:16:2 -> out=44100:32:2

**Comment by MaxKellermann** [comment]
Yes. This is not strictly bit-perfect, but "bit-perfect enough", because the conversion from 16 to 32 bit does not lose any precision.

**Reply by timsofteng** [reply]
Got it. Thank you.
So it is already not strictly bitperfect. May I somehow added crossfeed plugin into stream without dramatic losses?
